src/lavinmq/mqtt/broker.cr

```diff
@@ -70,14 +70,20 @@ module LavinMQ
         @vhost.rm_connection(client)
       end
 
-      def publish(packet : MQTT::Publish)
+      def publish(user : Auth::User, packet : MQTT::Publish)
+        unless user.can_write?(@vhost.name, packet.topic)
```

Comment:
This is wrong usage of `#can_write?`. The second argument should be the "resource name", in this case the exchange name.